### PR TITLE
port to node

### PR DESCRIPTION
**Pull request by arve0**
This is not done, but I'm opening this to get feedback.

What needs to be done before merging:

- [x] merge latest changes in master
- [ ] assure all blocks render correctly: `npm test`
- [x] package `translations.js` and `all-translations.js`
- [x] discuss API change, major version bump
- [x] document API change

Quite a few changes here, just leave a line comment if something is unclear or you disagree in the change.

### CODE REVIEW

File: src/fetch_translations.js [hunk 1]
```diff
@@ -42,6 +49,26 @@ var UNTRANSLATED = [
...
+  'if %b',
+  'forever if %b',
+  'stop script',
+  'stop all',
+  'switch to costume %m.costume',
+  'next background',
+  'switch to background %m.backdrop',
+  'background #',
+  'loud?',
   'username'
```
Comment (tjvr) [inline src/fetch_translations.js]: `username` should probably be required now!

**Comment by tjvr** [comment]
Hi! There's lots of great stuff here :D

Here are some thoughts:

- Rewriting `build_translations` in JS is excellent, I've been meaning to do that for a while.
- Porting `extra_aliases` to JSON is also rather sensible, thanks for that.
- Regarding the API change: I like the old API, and I'd rather not change it at this point (it's already deployed eg. on the forums, and the forums are very fragile.) Maybe add a convenience function (like `scratchblocks.renderToSVG(text)` if that's what you want? :)
- I'm uncertain about the build system changes. It's particularly important to me that I can hack on scratchblocks without having to manually rebuild after each change; my workflow is that I just refresh the browser window. Can you confirm that's still possible?
- I liked that scratchblocks was entirely self-contained in one static JS file, but I guess I'll give that up if I have to. (Hence the `Makefile` for updating its contents; I guess you've removed that now?)
- I'm not sure why you moved `utils.js` to a separate file. I'd rather keep everything together if possible 
:)
- The text measuring stuff is clever, but I'm not sure how well it will work for eg. unicode/emoji. I need to have a closer read through your code (the diff is pretty large because of all the file moves around, it'll take me a while to sort through).

If you can support headless rendering with fewer changes to the build system, that'd be great :) Otherwise this looks like a good piece of work!

**Comment by arve0** [comment]
> Regarding the API change: I like the old API, and I'd rather not change it at this point (it's already deployed eg. on the forums, and the forums are very fragile.) Maybe add a convenience function (like scratchblocks.renderToSVG(text) if that's what you want? :)

I understand that.

> I'm uncertain about the build system changes. It's particularly important to me that I can hack on scratchblocks without having to manually rebuild after each change; my workflow is that I just refresh the browser window. Can you confirm that's still possible?

Yes, it is. See `CONTRIBUTING.md` for work flow. Essentially this should set you up:

```
npm install
npm start
```

It does not livereload, but it does build when file changes.

> I liked that scratchblocks was entirely self-contained in one static JS file, but I guess I'll give that up if I have to.

Heh, I though of separating it even more. Had some hard time navigating through the flow of the code.

> Hence the Makefile for updating its contents; I guess you've removed that now?

Yes, all of that is contained in `package.json`, see `scripts`. It's basically equivalent as the makefile, but idiomatic node.

> I'm not sure why you moved utils.js to a separate file. I'd rather keep everything together if possible :)

I moved out code which would not run / give sense on node. I can put it back if you prefer, or maybe add it through the browser build step? Is it that it remains in a separate file that the problem, or that it's not available directly?

> The text measuring stuff is clever, but I'm not sure how well it will work for eg. unicode/emoji. I need to have a closer read through your code (the diff is pretty large because of all the file moves around, it'll take me a while to sort through).

Canvas should be equivalent to rendering in DOM. Here is the essential part:

```js
function measureText(text, font) {
     // re-use canvas object for better performance
     var canvas = measureText.canvas || (measureText.canvas = newCanvas());
     var context = canvas.getContext("2d");
     context.font = font;
     return context.measureText(text);
 }
```

**Comment by tjvr** [comment]
> Canvas should be equivalent to rendering in DOM. Here is the essential part:

Oh, great. I was thinking of your [proof of concept](https://github.com/tjvr/scratchblocks/pull/187/commits/d227bd709203fb86c9b5901a91e21649c1c2219c#diff-9d1343e9a0db0adaf30fe09f594dbb79R1) but presumably that's removed now. :)

> I can put it back if you prefer, or maybe add it through the browser build step? Is it that the code remains in a separate file that's the problem, or that the API is not available directly?

I'd prefer if it went back, I think. I'd like the old API put back and to be available; having it in one preference is a style thing.

> npm start

Should've said; I did try this, and it seemed to work the way I wanted -- I just wanted you to reassure me that's the intention! :)

I'm sorry to hear you had trouble understanding the code! Sounds like something a long comment / dev doc / philosophy page might help fix? I should probably try writing one!

Thanks again for working on this!

**Comment by arve0** [comment]
> Oh, great. I was thinking of your proof of concept but presumably that's removed now. :)

Hehe, yes, that was a naive dead end. What really amazed me was that there is so little concrete information about fonts. Like, what available font sizes are available, which fonts looks alike, etc. Did find this great resource though: [fontfamily.io](http://fontfamily.io/Lucida_Grande,Lucida_Sans_Unicode,Verdana,Arial,sans-serif;)

> I'm sorry to hear you had trouble understanding the code! Sounds like something a long comment / dev doc / philosophy page might help fix? I should probably try writing one!

To put it in context, first, I tried reading through. As it's a couple of thousand lines, keeping things in memory / context did not work out. 

Then I used the debugger, and drew a flow map. This way I was able to sort out essential parts. Still, I do not have complete overview of the code yet.

A philosophy / writeup would probably help here. Here are a couple of questions I still haven't answered:

- Is the code structured with a tokenizer, lexer and parser?
- What does the tokenized stream look like?

**Comment by tjvr** [comment]
> Is the code structured with a tokenizer, lexer and parser?
> What does the tokenized stream look like?

Ahaha, no! Parsing scratchblocks is really hard: it doesn't really fit the traditional lexer/parser model.

What we have is two separate parsers: `parseLines` and `parseScripts`. parseLines() is a recursive descent parser that operates over a stream of characters (the original source). It parses a whole line to get a `Block` --  or the string `'NL'` for blank lines. It takes care of things like auto-closing brackets at the end of a line, and it calls into `paintBlock` to do the block recognition.

parseScripts() streams in the lines and "parses" them to build up the nested structure -- C blocks, scripts, etc. It takes care of things like: a hat block always forces a new script.

`recogniseStuff` walks over the AST looking for things like list names and custom blocks, and recolours them appropriately.

Where this gets confusing is that in order to implement the `{ }` notation for rings / custom C blocks / embedded stack blocks, `parseScripts` also calls into `parseLines`! So the flow is somewhat inherently confusing :P

---

I'll add a philosophy writeup to my todo list; that'll go on my blog in a month or two. But if you have any more questions at any point, feel free to ask!

Are you happy to clean this up a bit, and restore the previous API? Then I can try and get this merged. :)

**Comment by tjvr** [comment]
Regarding the src/browser `scratchblocks.js` distinction -- would it make sense to use AMD/UMD here, and have a single file shared between both?

**Comment by arve0** [comment]
It should already be UMD:

```
  --standalone -s  Generate a UMD bundle for the supplied export name.
                   This bundle works with other module systems and sets the name
                   given as a window global if no module system is found.
```

**Comment by arve0** [comment]
>Are you happy to clean this up a bit, and restore the previous API?

I'll take a look at restoring the API now.

**Comment by arve0** [comment]
PS: You need to get the package name back on npmjs: https://www.npmjs.com/package/scratchblocks

**Comment by tjvr** [comment]
> PS: You need to get the package name transferred on npmjs: https://www.npmjs.com/package/scratchblocks

Yeah, I don't know how to go about that. Do you want to try taking it over? :)

**Comment by arve0** [comment]
Ok, this is good to go. Please take a look in the tests folder after a `npm test` to confirm that blocks renders correctly.

Also, I'm not including the latest changes in master, as more commits have been pushed just after I pull and merged the latest changes. I'm sure you can figure this out, I did `git checkout master; git diff v3.1` to find latest changes.

**Comment by tjvr** [comment]
> I'm not including the latest changes in master, as more commits have been pushed just after I pulled and merged the latest changes

Huh? I haven't pushed to master for at least a week (the last commit was eight days ago).

I don't like the style changes; for example, the 'inconsistent quotes' are actually a personal style thing (double-quotes for user-facing strings, generally). I'd also much rather the indentation in `src/scratchblocks.js` wasn't changed; this is making the diffs overly complicated.

**Comment by arve0** [comment]
> Huh? I haven't pushed to master for at least a week (the last commit was eight days ago).

See my edit above. Git did not figure out the manual merge, I misunderstood that.

> I don't like the style changes; for example, the 'inconsistent quotes' are actually a personal style thing (double-quotes for user-facing strings, generally).

I get that.

>  I'd also much rather the indentation in src/scratchblocks.js wasn't changed; this is making the diffs overly complicated.

Sorry about that.

I've already used quite some time here, and I'm leaving it with this. You can probably figure out how to keep the style of the project as you like.

**Comment by tjvr** [comment]
Ok, will do. :)

**Comment by arve0** [comment]
For people interested in node/UMD/browserify/webpack compatibility, I've forked to [arve0/scratchblocks](https://github.com/arve0/scratchblocks).

**Comment by tjvr** [comment]
Sorry—I keep meaning to get to this, but I'm just too busy!

I've merged your branch—from the small amount I've played with it, it looks very good :-)

I've also granted you push access, so you can merge your own pull requests in future.

**Pull request merged**